mesonbuild/interpreter/interpreter.py

```diff
@@ -2079,11 +2080,30 @@ def func_install_headers(self, node: mparser.BaseNode,
             if os.path.isabs(install_subdir):
                 mlog.deprecation('Subdir keyword must not be an absolute path. This will be a hard error in the next release.')
 
-        h = build.Headers(source_files, install_subdir, kwargs['install_dir'],
-                          kwargs['install_mode'], self.subproject)
-        self.build.headers.append(h)
+        if kwargs['preserve_path']:
+            dirs = {}
+            ret_headers = []
+            for file in source_files:
+                dirname = os.path.dirname(file.fname)
+                dirfiles = dirs.get(dirname)
+                if dirfiles is None:
+                    dirs[dirname] = [file]
+                else:
+                    dirfiles.append(file)
```

Comment:
This would be more elegantly written if `dirs = collections.defaultdict(list)`, then you can simply append to it without regard to whether it exists.